Add --help and tab completion to rbenv-versions

jasonkarns opened the pull request:
- adds --help option for full help
- adds tab completion (--bare, --skip-aliases)

Questions:
- should --help itself be tab completed?
- should --help option be listed in usage?

jasonkarns commented:
Given answers to the above questions, I'd like to extend this PR to make the tab completion for other subcommands consistent.

mislav commented on libexec/rbenv-versions:
```diff
@@ -11,7 +11,13 @@ unset bare
   case "$arg" in
+  --complete )
+    echo --bare
+    echo --help
```
Before you add `--help` to every command, we should maybe consider solving that on another level. Right now the preferred way to get help was `rbenv help <command>`. Maybe we could change the router to make `rbenv <command> --help` do the same thing? That would save us from having to edit every subcommand's argument parsing.

jasonkarns replied:
I like this idea. I'm mainly modeling this on git, which supports both `git help <command>` and `git <command> --help` (and the fact that some plugins/commands already had --help but others didn't.)

jasonkarns replied:
We could also update the tab completion portion to always echo `--help` as well (before invoking the subcommand's `--complete`).

mislav replied:
I guess so, yeah. I don't find that to be extremely valuable, but for the sake of correctness.

mislav commented on libexec/rbenv-versions:
```diff
@@ -11,7 +11,13 @@ unset bare
 unset skip_aliases
 for arg; do
   case "$arg" in
+  --complete )
```
When you add this, you also have to add this line somewhere:

```
# provide rbenv completions
```

It's not intuitive, but this magic comment "marks" the command as being completion-aware.

jasonkarns replied:
right, not sure how I missed that in this one. I definitely had it at one point, because I verified it was working :/

jasonkarns commented:
With the general `--help` feature merged, I've rebased this branch.

- added completions for `completions`, `init`, and `versions`
- made `help`'s completions consistent with others

After this pr, all commands under libexec have respective completions.

mislav commented on libexec/rbenv-completions:
```diff
@@ -10,6 +10,11 @@ if [ -z "$COMMAND" ]; then
 
+# Provide rbenv completions
+if [ "$COMMAND" = "--complete" ]; then
+  exec rbenv-commands
```
Do you see a value in providing completion support for `rbenv-completion`? After all, people shouldn't generally use this command directly in their interactive shell.

jasonkarns replied:
I'm so-so on it. Most people probably don't, but I use it a lot when monkeying with rbenv/nodenv. So marginal usefulness + consistency + no downsides = enough to convince me. :)

mislav commented:
Generally looks :

jasonkarns commented:
I was more unconvinced on the inclusion of the `-` in rbenv-init. But since it's virtually a required argument, I decided to keep it in. :shrug:
